crates/core/src/db/datastore/traits.rs

```diff
@@ -158,29 +159,49 @@ pub enum IsolationLevel {
     Serializable,
 }
 
-/// Operations in a transaction are either Inserts or Deletes.
-/// Inserts report the byte objects they inserted, to be persisted
-/// later in an object store.
-pub enum TxOp {
-    Insert(Arc<Vec<u8>>),
-    Delete,
+/// A record of all the operations within a transaction.
+#[derive(Default)]
+pub struct TxData {
+    /// The inserted rows per table.
+    inserts: BTreeMap<TableId, Arc<[ProductValue]>>,
+    /// The deleted rows per table.
+    deletes: BTreeMap<TableId, Arc<[ProductValue]>>,
+    /// Map of all `TableId`s in both `inserts` and `deletes` to their
+    /// corresponding table name.
+    tables: BTreeMap<TableId, String>,
```

Comment:
Pure curiosity: Why `BTreeMap` instead of `HashMap`?

Reply:
It is significant when replaying that system tables (lower ids) come first, so creating and mutating a table in the same transaction works.

The way the visitor pattern reading works, this also means keeping them ordered on disk. This may be debatable…

Reply:
I changed `tables` to use `spacetimedb_data_structures::map::IntMap`, which is also used to build the `DatabaseUpdate`.

Or did you indeed mean `IndexMap` (i.e. insertion ordered)?

Reply:
Does `IntMap` provide the ordering guarantees you need?

Reply:
So `tables` doesn’t need ordering at all, and we “know” table ids are unique.

The ordering of inserts and deletes is significant, although if we _somehow_ restrict construction to the `merge` function, we’d also have them ordered already.

I just don’t want to docstring-type that, as I’ve had many issues in the past with implied ordering / set semantics.